### ISSUE TITLE
swap_view does not work Hosted Feature Services with multiple layers

### ISSUE THREAD

**Issue by stevewayuk** [Esri/arcgis-python-api#2029]
**Describe the bug**
The swap_view function allows the ability to swap the source of the view to a different Hosted Feature Layer.

https://developers.arcgis.com/python/api-reference/arcgis.features.managers.html#arcgis.features.managers.FeatureLayerCollectionManager.swap_view

Hosted Feature Layer Views can only be configured against a single service.

When using the swap_view function, you pass the layer index, as well as the new layer source.  The current function will never work where more than one layer exists in the view, as soon as you try and change one layer, an error is returned saying the view does not support multiple services, this is because the function requires a layer index.

**To Reproduce**

1) Create a Hosted Feature Layer View against a service with multiple layers.  Ensure the view has multiple layers.

Assume 2 layers exist (0,1).

2) execute the function: 

`swap_view(0,"https://services-eu1.arcgis.com/xyz/arcgis/rest/services/ServiceName/FeatureServer/0")`

error:
```
--- Swapping View Layer Source ---
Unable to add feature service definition.
View layers can reference only one source service (AIMSAidtoNavigation_20240904_COPY).
(Error Code: 400)
 
 
--- Swapping View Layer Source ---
Unable to delete feature service definition.
Object reference not set to an instance of an object.
(Error Code: 400)
```

**Expected behavior**
It is expected that this function should be able to provide a mechanism to swap a view with multiple layers.  I expect this is as designed, but it currently not very useful.

**Platform (please complete the following information):**
 - OS: Windows
 - Python API Version [e.g. `2.3.1`]

**Comment by achapkowski** [comment]
You are passing in a URL, it takes  Table or FeatureLayer

**Comment by stevewayuk** [comment]
> You are passing in a URL, it takes Table or FeatureLayer

I've tried both ways.....it still provides the same error.

It is because when you specify a layer index, then at some point the view would have layers from multiple services.

**Comment by Robert-Holliday** [comment]
I also am having this problem. Has there been any headway towards a suggested solution?

**Comment by achapkowski** [comment]
Can you post your code?

**Comment by Robert-Holliday** [comment]
[swap_view_dev.zip](https://github.com/user-attachments/files/17938628/swap_view_dev.zip)

**Comment by Robert-Holliday** [comment]
Done. Thanks for looking.


Bob Holliday, GIS Analyst
    ***@***.***  ***@***.******@***.***>
    ***@***.***  406.444.0770
     ***@***.***  msl.mt.gov<http://www.msl.mt.gov/>

***@***.***<http://www.msl.mt.gov/>
 _________________________



From: Andrew ***@***.***>
Sent: Wednesday, November 27, 2024 10:15 AM
To: Esri/arcgis-python-api ***@***.***>
Cc: Holliday, Robert ***@***.***>; Comment ***@***.***>
Subject: [EXTERNAL] Re: [Esri/arcgis-python-api] swap_view does not work Hosted Feature Services with multiple layers (Issue #2029)


Can you post your code?

**Comment by ajs902** [comment]
Bumping this as it is making the function unusable.

**Comment by achapkowski** [comment]
@Robert-Holliday what version of the API are you seeing this in?  Example 2.4.0? 

I see you have
```python
gis = GIS("home")
```

So I can't infer what platform you are running on, can you please let me know if this is enterprise + version or online.

**Comment by ajs902** [comment]
AGOL 2.4.2

**Comment by Robert-Holliday** [comment]
> AGOL 2.4.2

Correct, AGOL in my case, but I haven't verified functionality @ 2.4.2.

This issue will be addressed in a future release. I worked through a support case and then went through the enhancement request process. Then late last month I got the following message from ESRI.

ArcGIS product teams are preparing to address ENH-000172892

Description: Extend the capabilities of the swap_view method in the ArcGIS API for Python to include swapping the source of views with more than one layer

The product teams have prioritized this issue for implementation in a future release. Typically, once an issue enters this status, it is implemented within one to two release cycles. At this stage, software issues are still subject to change and could be dropped or postponed. You will be notified of future updates. 

Esri Inc.
THE SCIENCE OF WHERE™

**Comment by Robert-Holliday** [comment]
Here is code I'm using for updating view sources


def swap_view_source(
        source_item: Item,
        target_service_name: str,
        portal_url: str,
        portal_content_owner: str,
        portal_password: str,
        portal_item_type: str,
        lf: TextIO
        ) -> None | Exception:
    """Updates an AGOL hosted wiew layer.

    Parameters:
        source_item: The AGOL item with updates.
        target_service_name: The name of the AGOL item to be updated.
        portal_url: The url of the portal.
        portal_content_owner: The portal account which owns the source item and the target item.
        portal_password: The password for portal content owner account.
        portal_item_type: The type of portal item to be processed.
        lf: Log file.
    Retuns:
        None: Successful completion
        Exception: Error encountered during execution
    """
    from arcgis import GIS
    from arcgis.features import FeatureLayerCollection
    import ast
    import concurrent.futures
    import os

    error_message: Any = None
    try:
        #
        #       Connect to Portal
        gis = None
        gis = GIS(
            url=portal_url,
            username=portal_content_owner,
            password=portal_password,
            verify_cert=True
        )
        portal_item: Item
        wc: str = f"title:{target_service_name}"
        target_view: Item
        source_flc: FeatureLayerCollection = FeatureLayerCollection.fromitem(source_item)
        for portal_item in [
            pi for pi in gis.content.search(
                query=wc, item_type=portal_item_type, max_items=10000
                )
            if pi.title == target_service_name
        ]:
            target_view: Item = gis.content.get(portal_item.id)
            target_view_flc: FeatureLayerCollection = FeatureLayerCollection.fromitem(target_view)
            #
            #       Unregister Replicas
            if target_view_flc.properties.syncEnabled:
                for rep in [r for r in target_view_flc.replicas.get_list()]:
                    message = f"....unregistering replica: {rep['replicaName']}"
                    lf.write(f"{message}\n")
                    print(message)
                    a_replica = target_view_flc.replicas.get(rep['replicaID'])
                    target_view_flc.replicas.unregister(a_replica['replicaID'])
            #
            #       Disable Sync
            futures: list = [
                target_view_flc.manager.update_definition({'syncEnabled': False}, future=True)
            ]

            # Wait for futures to complete
            concurrent.futures.wait(futures)
            message = "....Updating view source"
            lf.write(f"{message}\n")
            print(message)
            #       swap_view if target view contains only one layer
            for a_layer in [
                lyr for lyr in target_view_flc.layers
                if len(target_view_flc.layers) == 1
            ]:
                layer_index: int = int(a_layer.properties.id)
                target_view_flc.manager.swap_view(
                    layer_index, source_item.layers[layer_index]  # type: ignore
                    )
            #       Update layer definitions if target view contains more than one layer
            add_dict: dict = {'layers': []}
            delete_list: list = []
            source_service_name: str = os.path.basename(os.path.dirname(source_flc.url))
            for a_layer in [
                lyr for lyr in source_flc.layers
                if len(target_view_flc.layers) > 1
            ]:
                layer_id: int = int(a_layer.properties.id)
                temp = ast.literal_eval(f"{'{'}'id': {layer_id}{'}'}")
                delete_list.append(temp)
                add_dict['layers'].append(
                    {
                        'adminLayerInfo': {
                            'geometryField': {'name': 'Shape'},
                            'xssTrustedFields': '',
                            'viewLayerDefinition':
                            {
                                'sourceServiceName': source_service_name,
                                'sourceLayerId': layer_id,
                                'sourceLayerFields':  '*'
                            }
                        },
                        'id': layer_id,
                        'name': a_layer.manager.properties['name'],
                        'drawingInfo': a_layer.manager.properties['drawingInfo']
                    }
                )
            delete_dict = dict({'layers': delete_list})
            futures: list = [
                target_view_flc.manager.delete_from_definition(delete_dict, future=True)
            ]
            # Wait for futures to complete
            concurrent.futures.wait(futures)
            futures: list = [
                target_view_flc.manager.add_to_definition(add_dict, future=True)
            ]
            # Wait for futures to complete
            concurrent.futures.wait(futures)
            #
            #       Update service properties
            message = "....Updating cache timeout, export, and sync"
            lf.write(f"{message}\n")
            print(message)
            futures: list = [
                #       Update capabilities
                target_view_flc.manager.update_definition(
                    {'capabilities': 'Query,Extract'}, future=True
                    ),
                #       Set cache max age to 1 hour
                target_view_flc.manager.update_definition({'cacheMaxAge': 3600}, future=True),
                #       Enable sync
                target_view_flc.manager.update_definition({'syncEnabled': True}, future=True)
            ]

            #       Wait for futures to complete
            concurrent.futures.wait(futures)

        if not target_view:
            error_message = f"The hosted view layer {target_service_name} does not exist"
            raise Exception(error_message)
        return None

    except Exception as e:
        return e

**Comment by nanaeaubry** [comment]
This has been fixed for version 2.4.3 of the API
